core/src/main/scala/com/microsoft/azure/synapse/ml/causal/SyntheticDiffInDiffEstimator.scala

```diff
@@ -0,0 +1,147 @@
+
+package com.microsoft.azure.synapse.ml.causal
+
+import com.microsoft.azure.synapse.ml.codegen.Wrappable
+import com.microsoft.azure.synapse.ml.core.schema.DatasetExtensions
+import com.microsoft.azure.synapse.ml.logging.FeatureNames
+import org.apache.spark.ml.param.{DoubleParam, ParamValidators, Params}
+import org.apache.spark.ml.{ComplexParamsReadable, ComplexParamsWritable}
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.{BooleanType, IntegerType}
+import org.apache.spark.sql.{Dataset, Row}
+
+trait SyntheticDiffInDiffEstimatorParams extends SyntheticEstimatorParams {
+  final val zeta = new DoubleParam(this, "zeta",
+    "The zeta value for regularization term when fitting unit weights. " +
+      "If not specified, a default value will be computed based on formula (2.2) specified in " +
+      "https://www.nber.org/system/files/working_papers/w25532/w25532.pdf. " +
+      "For large scale data, one may want to tune the zeta value, minimizing the loss of the unit weights regression.",
+    ParamValidators.gtEq(0))
+
+  def getZeta: Double = $(zeta)
+  def setZeta(value: Double): this.type = set(zeta, value)
+}
+
+class SyntheticDiffInDiffEstimator(override val uid: String)
+  extends BaseDiffInDiffEstimator(uid)
+    with SyntheticEstimator
+    with SyntheticDiffInDiffEstimatorParams
+    with ComplexParamsWritable
+    with Wrappable {
+
+  logClass(FeatureNames.Causal)
+
+  import SyntheticEstimator._
+
+  def this() = this(Identifiable.randomUID("SyntheticDiffInDiffEstimator"))
+
+  // scalastyle:off method.length
+  override def fit(dataset: Dataset[_]): DiffInDiffModel = logFit({
+    val df = dataset
+      .withColumn(getTreatmentCol, treatment.cast(BooleanType))
+      .withColumn(getPostTreatmentCol, postTreatment.cast(BooleanType))
+      .toDF
+    val controlDf = df.filter(not(treatment)).cache
+    val preDf = df.filter(not(postTreatment)).cache
+    val timeIdxCol = DatasetExtensions.findUnusedColumnName("Time_idx", df)
+    val unitIdxCol = DatasetExtensions.findUnusedColumnName("Unit_idx", df)
+    val timeIdx = createIndex(preDf, getTimeCol, timeIdxCol).cache
+    val unitIdx = createIndex(controlDf, getUnitCol, unitIdxCol).cache
+    val size = (unitIdx.count, timeIdx.count)
+
+    // indexing
+    val indexedControlDf = controlDf.join(timeIdx, controlDf(getTimeCol) === timeIdx(getTimeCol), "left_outer")
+      .join(unitIdx, controlDf(getUnitCol) === unitIdx(getUnitCol), "left_outer")
+      .select(unitIdxCol, timeIdxCol, getTreatmentCol, getPostTreatmentCol, getOutcomeCol)
+      .localCheckpoint(true)
+
+    val indexedPreDf = preDf.join(timeIdx, preDf(getTimeCol) === timeIdx(getTimeCol), "left_outer")
+      .join(unitIdx, preDf(getUnitCol) === unitIdx(getUnitCol), "left_outer")
+      .select(unitIdxCol, timeIdxCol, getTreatmentCol, getPostTreatmentCol, getOutcomeCol)
+      .localCheckpoint(true)
```

Comment:
These outer joins were an issue in shap, as the way scott L implemented shap was to randomly sample a random bg sample instead of doing the full outer join. Im not sure if this comment makes any sense here, but its food for thought to ask what happens when the data gets very large here. Maybe it doesent matter and this is necessary for the algo, just wanted to double check.

Reply:
So the problem with the SHAP library is because I did a **cross join** there, which causes data explosion - I'll look into that problem after I investigate the ONNX-related issues first. Here we're performing left outer joins, and the `timeIdx` dataframe is guaranteed to be unique (see `createIndex` function, it calls `assignRowIndex` function, which relies on rdd.zipWithIndex) on the join key, so it wouldn't cause data explosion.